crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,599 @@
+    })
+    .pipe(Ok)
+}
+
+/// Query parameters for the resolve transaction endpoint
+#[derive(Debug, Default, serde::Serialize, serde::Deserialize, JsonSchema)]
+pub struct ResolveTransactionQueryParameters {
+    /// Request that the fully resolved transaction be simulated and have its results sent back in
+    /// the response.
+    #[serde(default)]
+    pub simulate: bool,
+    #[serde(flatten)]
+    pub simulate_transaction_parameters: SimulateTransactionQueryParameters,
+}
+
+struct NormalizedPackage {
+    #[allow(unused)]
+    package: MovePackage,
+    normalized_modules: BTreeMap<String, normalized::Module>,
+}
+
+fn called_packages(
+    reader: &StateReader,
+    protocol_config: &ProtocolConfig,
+    unresolved_transaction: &UnresolvedTransaction,
+) -> Result<HashMap<ObjectId, NormalizedPackage>> {
+    let binary_config = sui_types::execution_config_utils::to_binary_config(protocol_config);
+    let mut packages = HashMap::new();
+
+    for move_call in unresolved_transaction
+        .ptb
+        .commands
+        .iter()
+        .filter_map(|command| {
+            if let Command::MoveCall(move_call) = command {
+                Some(move_call)
+            } else {
+                None
+            }
+        })
+    {
+        let package = reader
+            .inner()
+            .get_object(&(move_call.package.into()))?
+            .ok_or_else(|| ObjectNotFoundError::new(move_call.package))?
+            .data
+            .try_as_package()
+            .ok_or_else(|| {
+                RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    format!("object {} is not a package", move_call.package),
+                )
+            })?
+            .to_owned();
+
+        // Normalization doesn't take the linkage or type origin tables into account, which means
+        // that if you have an upgraded package that introduces a new type, then that type's
+        // package ID is going to appear incorrectly if you fetch it from its normalized module.
+        //
+        // Despite the above this is safe given we are only using the signature information (and in
+        // particular the reference kind) from the normalized package.
+        let normalized_modules = package.normalize(&binary_config).map_err(|e| {
+            RestError::new(
+                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
+                format!("unable to normalize package {}: {e}", move_call.package),
+            )
+        })?;
+        let package = NormalizedPackage {
+            package,
+            normalized_modules,
+        };
+
+        packages.insert(move_call.package, package);
+    }
+
+    Ok(packages)
+}
+
+fn resolve_unresolved_transaction(
+    reader: &StateReader,
+    called_packages: &HashMap<ObjectId, NormalizedPackage>,
+    reference_gas_price: u64,
+    max_gas_budget: u64,
+    unresolved_transaction: UnresolvedTransaction,
+) -> Result<TransactionData> {
```

Comment:
Noob question: could we have an API that returns `sui-sdk-types::types::Transaction` as well, or what'd be the best way to convert this `TransactionData` into  a `Transaction`, just bcs-serialize-deserialize?

Reply:
this function is an internal implementation detail so i wouldn't worry about the type here. If you look at the REST entry point you'll find that it does return the the sdk types version of things

Reply:
Yes, that's why I was a bit confused 😅